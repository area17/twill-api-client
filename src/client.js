import { stringify } from 'qs'
import { ofetch } from 'ofetch'

export const client = () => {
  return async (url, fetchOptions = {}) => {
    const headers = {
      ...(fetchOptions.headers || {})
    }

    if (fetchOptions.params) {
      url = `${url}${url.includes('?') ? '&' : '?'}${stringify(
        fetchOptions.params
      )}`
      delete fetchOptions.params
    }

    return await ofetch(url, {
      responseType: 'json',
      retry: 0,
      ...fetchOptions,
      headers
    }).catch((error) => {
      throw error?.data?.errors || error
    })
  }
}
