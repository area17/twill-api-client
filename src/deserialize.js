import { camelCase, isEmpty } from 'lodash-es'
import { camelCaseKeys } from './utils/camel-case-keys'

const MAX_DEPTH = 6

function deserializeRelationships(resources = [], store, depth) {
  return resources
    .map((resource) => deserializeRelationship(resource, store, depth))
    .filter((resource) => !!resource)
}

function deserializeRelationship(resource = {}, store, depth) {
  if (
    typeof resource === 'object' &&
    resource?.type &&
    store[camelCase(resource.type)] &&
    store[camelCase(resource.type)][resource.id]
  ) {
    return deserialize(
      store[camelCase(resource.type)][resource.id],
      store,
      depth
    )
  }

  return null
}

export function deserialize(result, store, depth) {
  if (Array.isArray(result)) {
    return result.map((item) => deserialize(item, store, depth))
  }

  const { id, type, attributes, relationships, meta } =
    store[camelCase(result.type)][result.id]

  if (typeof depth == 'number') {
    depth++
  } else {
    depth = 1
  }

  /**
   * If max depth is reached, simply return the serialized resource
   */
  if (depth > MAX_DEPTH) {
    return { id, type, attributes, relationships, meta }
  }

  let resource = {
    type,
    meta,
    id,
    ...camelCaseKeys(attributes)
  }

  if (relationships) {
    try {
      resource = Object.keys(relationships).reduce((resource, key) => {
        return {
          ...resource,
          [camelCase(key)]: !isEmpty(relationships[key].data)
            ? Array.isArray(relationships[key].data)
              ? deserializeRelationships(relationships[key].data, store, depth)
              : deserializeRelationship(relationships[key].data, store, depth)
            : relationships[key].data
        }
      }, resource)
    } catch (err) {
      console.log(err)
    }
  }

  return resource
}
