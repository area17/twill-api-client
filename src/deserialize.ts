import { camelCase, isEmpty } from 'lodash-es'
import {
  Resource,
  ID,
  JsonApiRelatedResource,
  JsonApiResource,
  NormalizedStore,
} from '@/types'
import { camelCaseKeys } from '@/utils/camel-case-keys'

const MAX_DEPTH = 6

function deserializeRelationships(
  resources: JsonApiRelatedResource[],
  store: NormalizedStore,
  depth: number,
): Resource[] {
  return resources
    .map((resource) => deserializeRelationship(resource, store, depth))
    .filter((resource) => !!resource)
}

function deserializeRelationship(
  resource: JsonApiRelatedResource,
  store: NormalizedStore,
  depth: number,
): Resource {
  return deserialize(resource, store, depth) as Resource
}

export function deserialize<Type extends { type: string; id: ID }>(
  result: Type[] | Type,
  store: NormalizedStore,
  depth = 0,
): Resource | Resource[] {
  if (Array.isArray(result)) {
    return result.map((result) => deserialize(result, store, depth) as Resource)
  }

  const serializedResource: JsonApiResource =
    store[camelCase(result.type)][result.id]

  if (typeof depth == 'number') {
    depth++
  } else {
    depth = 1
  }

  /**
   * If max depth is reached, simply return the serialized resource
   */
  if (depth > MAX_DEPTH) {
    return { ...serializedResource } as Resource
  }

  let resource: Resource = {
    type: serializedResource.type,
    meta: serializedResource.meta,
    id: serializedResource.id,
    ...camelCaseKeys(serializedResource.attributes),
  }

  if (serializedResource.relationships) {
    resource = Object.keys(serializedResource.relationships).reduce(
      (acc, key) => {
        const data = serializedResource.relationships[key].data
        let relationship

        if (!isEmpty(data)) {
          relationship = Array.isArray(data)
            ? deserializeRelationships(data, store, depth)
            : deserializeRelationship(data, store, depth)
        }

        return {
          ...acc,
          [camelCase(key)]: relationship ?? data,
        }
      },
      resource,
    )
  }

  return resource
}
