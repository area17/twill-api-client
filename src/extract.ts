import { Resource, Transformations } from '@/types'
import { blocks } from '@/extract/blocks'
import { browsers } from '@/extract/related-items'
import { files } from '@/extract/files'
import { media } from '@/extract/media'

export const extract = (resource: Resource) => {
  const extracted: Transformations = {}

  if (Array.isArray(resource.blocks)) {
    extracted.blocks = blocks(resource)
  }

  if (Array.isArray(resource.media)) {
    extracted.media = media(resource)
  }

  if (Array.isArray(resource.relatedItems)) {
    extracted.relatedItems = browsers(resource)
  }

  if (Array.isArray(resource.files)) {
    extracted.files = files(resource)
  }

  return extracted
}
