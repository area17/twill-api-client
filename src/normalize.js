import { camelCase } from 'lodash-es';
import { camelCaseKeys } from '@/utils/camelCaseKeys';

/**
 * Normalize the JSON:API response into categories by resource type
 * and indexed by the resource's ID.
 */
export function normalize(response) {
  let data;

  if (Array.isArray(response.data)) {
    data = response.data;
  } else {
    data = [response.data];
  }

  const included = response.included || [];

  let result = [];
  let resources = {};

  data.forEach((resource) => {
    addResult(result, resource);
    addResource(resources, resource);
  });

  included.forEach((resource) => {
    addResource(resources, resource);
  });

  return {
    result,
    resources
  };
}

function addResult(result, resource) {
  const { type, id } = resource;

  result.push({ type, id });
}

function addResource(resources, resource) {
  const { type, id, attributes, meta, links, relationships } = resource;

  const resourceType = camelCase(type);

  if (!resources[resourceType]) resources[resourceType] = {};

  resources[resourceType][id] = {
    id,
    type,
    attributes: camelCaseKeys(attributes || {}),
    relationships: extractRelationships(relationships),
    meta,
    links
  };

  return resources;
}

function extractRelationships(responseRelationships) {
  if (!responseRelationships) {
    return undefined;
  }

  let relationships = {};

  Object.keys(responseRelationships).map((type) => {
    const relationshipType = camelCase(type);
    relationships[relationshipType] = responseRelationships[type];
  });

  return relationships;
}
