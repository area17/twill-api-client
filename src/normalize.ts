import { camelCase } from 'lodash-es'
import { camelCaseKeys } from './utils/camel-case-keys'
import {
  NormalizedResult,
  NormalizedStore,
  JsonApiRelationship,
  JsonApiDataResponse,
  JsonApiResource,
  JsonApiRelatedResource,
} from './types'

/**
 * Normalize the JSON:API response into categories by resource type
 * and indexed by the resource's ID.
 */
export function normalize(response: JsonApiDataResponse): {
  result: NormalizedResult[]
  resources: NormalizedStore
} {
  let data

  if (Array.isArray(response.data)) {
    data = response.data
  } else {
    data = [response.data]
  }

  const included = response.included || []

  const result: NormalizedResult[] = []
  const resources: NormalizedStore = {}

  data.forEach((resource) => {
    addResult(result, resource)
    addResource(resources, resource)
  })

  included.forEach((resource) => {
    addResource(resources, resource)
  })

  return {
    result,
    resources,
  }
}

function addResult(result: NormalizedResult[], resource: JsonApiResource) {
  const { type, id } = resource

  result.push({ type, id })
}

function addResource(resources: NormalizedStore, resource: JsonApiResource) {
  const { type, id, attributes, meta, links, relationships } = resource

  const resourceType = camelCase(type)

  if (!resources[resourceType]) resources[resourceType] = {}

  resources[resourceType][id] = {
    id,
    type,
    attributes: camelCaseKeys(attributes || {}),
    relationships: extractRelationships(relationships) || {},
    meta,
    links,
  }

  return resources
}

function extractRelationships(responseRelationships: {
  [key: string]: JsonApiRelationship<JsonApiRelatedResource>
}): Record<string, JsonApiRelationship<JsonApiRelatedResource>> | undefined {
  if (!responseRelationships) {
    return undefined
  }

  const relationships: Record<
    string,
    JsonApiRelationship<JsonApiRelatedResource>
  > = {}

  Object.keys(responseRelationships).map((type) => {
    const relationshipType = camelCase(type)
    relationships[relationshipType] = responseRelationships[type]
  })

  return relationships
}
