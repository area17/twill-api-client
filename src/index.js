import { QueryBuilder } from './QueryBuilder.js'
import { normalize } from './normalize.js'
import { deserialize } from './deserialize'
import { allImages } from './extract/images.js'
import { allBlocks } from './extract/blocks.js'

const Twill = ({ url, prefix, version, token }) => {
  const baseURL = `${url}${prefix}/${version}`

  const headers = {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.api+json'
  }

  const get = (path) => {
    return new QueryBuilder({ path, headers })
  }

  const find = (resource) =>
    new QueryBuilder({
      path: resource,
      baseURL,
      headers
    })

  const findOne = (resource, id) =>
    new QueryBuilder({
      path: `${resource}/${id}`,
      baseURL,
      headers
    })

  const findRelated = (resource, response) => {
    const path = response.relationships[resource].links.related

    return new QueryBuilder({ path, headers })
  }

  const findRelationship = (resource, response) => {
    const path = response.relationships[resource].links.self

    return new QueryBuilder({ path, headers })
  }

  const extract = (resource) => {
    return {
      images: allImages(resource),
      editors: allBlocks(resource)
    }
  }

  const transform = (response) => {
    const normalized = normalize(response)
    const resources = deserialize(normalized.result, normalized.resources)
    return resources.map((resource) => {
      return {
        ...resource,
        ...extract(resource)
      }
    })
  }

  return {
    get,
    find,
    findOne,
    findRelated,
    findRelationship,
    normalize,
    deserialize,
    extract,
    transform
  }
}

export { Twill }
