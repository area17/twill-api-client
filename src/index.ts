import { QueryBuilder } from './query-builder'
import {
  ID,
  JsonApiDataResponse,
  JsonApiResource,
  DeserializedResource
} from './types'
import { normalize } from './normalize'
import { deserialize } from './deserialize'
import { blocks } from './extract/blocks'
// @ts-ignore
import { images } from './extract/images'

export interface TwillOptions {
  url: string
  token: string
  prefix: string
  version: string
}

export const Twill = (options: TwillOptions) => {
  const { url, prefix, version, token } = options
  const baseURL = `${url}${prefix}/${version}`

  const headers = {
    Authorization: `Bearer ${token}`,
    Accept: 'application/vnd.api+json'
  }

  const get = (path: string): QueryBuilder => {
    return new QueryBuilder({ path, headers })
  }

  const find = (resource: string) =>
    new QueryBuilder({
      path: resource,
      baseURL,
      headers
    })

  const findOne = (resource: string, id: ID | number | number) =>
    new QueryBuilder({
      path: `${resource}/${id}`,
      baseURL,
      headers
    })

  const findRelated = (
    resource: string,
    response: JsonApiResource
  ): QueryBuilder | null => {
    const path = response.relationships[resource]?.links?.related

    if (!path) {
      return null
    }

    return new QueryBuilder({ path, headers })
  }

  const findRelationship = (
    resource: string,
    response: JsonApiResource
  ): QueryBuilder | null => {
    const path = response.relationships[resource]?.links?.self

    if (!path) {
      return null
    }

    return new QueryBuilder({ path, headers })
  }

  const extract = (resource: DeserializedResource) => {
    return {
      images: images(resource),
      editors: blocks(resource)
    }
  }

  const transform = (response: JsonApiDataResponse) => {
    const normalized = normalize(response)
    const resources: DeserializedResource[] = deserialize(
      normalized.result,
      normalized.resources
    ) as DeserializedResource[]
    return resources.map((resource: DeserializedResource) => {
      return {
        ...resource,
        ...extract(resource)
      }
    })
  }

  return {
    get,
    find,
    findOne,
    findRelated,
    findRelationship,
    normalize,
    deserialize,
    extract,
    transform
  }
}
