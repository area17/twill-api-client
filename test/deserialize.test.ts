import {expect, test} from 'vitest'
import {normalize} from '../src/normalize'
import {deserialize} from '../src/deserialize'
import {data, error} from './response.test'

test('deserialize', () => {
  const normalized = normalize(data)
  const deserialized = deserialize(normalized.result, normalized.resources)
  expect(Array.isArray(deserialized)).toBe(true)
  expect(deserialized.length).toBe(1)
  expect(deserialized[0].attributes).toBeTypeOf('object')
  expect(deserialized[0].attributes.slug).toBe('slug')
  expect(deserialized[0].relationships).toBeTypeOf('object')
  expect(deserialized[0].relationships.blocks).toBeTypeOf('object')
  expect(deserialized[0].relationships.blocks.data).toHaveLength(1)
  expect(deserialized[0].relationships.blocks.links).toBeTypeOf('object')
  expect(deserialized).toMatchSnapshot()

  const blocks = deserialized[0].relationships.blocks
  expect(blocks.data[0].attributes.blockType).toBe('images')
})
