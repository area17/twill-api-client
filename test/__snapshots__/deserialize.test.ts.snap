// Vitest Snapshot v1

exports[`deserialize 1`] = `
[
  {
    "attributes": {
      "slug": "slug",
    },
    "id": "1",
    "links": {
      "self": "#",
    },
    "meta": undefined,
    "relationships": {
      "blocks": {
        "data": [
          {
            "attributes": {
              "blockType": "images",
              "position": 1,
            },
            "id": "1",
            "links": {
              "self": "#",
            },
            "meta": undefined,
            "relationships": {},
            "type": "blocks",
          },
        ],
        "links": {
          "related": "#",
          "self": "#",
        },
        "meta": {
          "editors": {
            "default": [
              "1",
            ],
          },
        },
      },
    },
    "type": "pages",
  },
]
`;
